pkg/datapath/linux/routing/migrate.go

```diff
@@ -0,0 +1,628 @@
+				failed++
+			} else {
+				if rs, found := cleanup[r]; found {
+					rs = append(rs, routes...)
+					cleanup[r] = rs
+				} else {
+					cleanup[r] = routes
+				}
+				migrated++
+			}
+		}
+	} else if isDowngrade {
+		for _, r := range v2Rules {
+			if routes, err := m.downgradeRule(r); err != nil {
+				log.WithError(err).WithField("rule", r).Warn("Failed to downgrade endpoint to original ENI datapath. " +
+					"Previous datapath is still intact and endpoint connectivity is not affected.")
+				failedTableIDs[r.Table] = struct{}{}
+				failed++
+			} else {
+				if rs, found := cleanup[r]; found {
+					rs = append(rs, routes...)
+					cleanup[r] = rs
+				} else {
+					cleanup[r] = routes
+				}
+				migrated++
+			}
+		}
+	}
+
+	// We store the routes that have already been deleted to de-duplicate and
+	// avoid netlink returning "no such process" for a route that has already
+	// been deleted. Note the map key is a string representation of a
+	// netlink.Route because netlink.Route is not a valid map key because it is
+	// incomparable due to containing a slice inside it.
+	deleted := make(map[string]struct{}, len(cleanup))
+
+	for rule, routes := range cleanup {
+		toDelete := make([]netlink.Route, 0, len(routes))
+		for _, ro := range routes {
+			if _, skip := failedTableIDs[rule.Table]; skip {
+				continue
+			}
+
+			if _, already := deleted[ro.String()]; !already {
+				// Declare the routes deleted here before the actual deletion
+				// below because we don't care if deletion succeeds or not. See
+				// comment below on why.
+				deleted[ro.String()] = struct{}{}
+				toDelete = append(toDelete, ro)
+			}
+		}
+
+		// This function does not return a revert stack unlike the others
+		// because this operation is best-effort. If we fail to delete old
+		// routes, then it simply means there is just leftover state left
+		// behind, but it has no impact on the datapath whatsoever. We can make
+		// that assumption because by the time we call this function, we'd have
+		// successfully deleted the old rule which would steer traffic towards
+		// these routes.
+		//
+		// We also don't want to revert here because at this point, the new
+		// datapath is in-place and it wouldn't make sense to risk reverting in
+		// case of a failure, just to merely cleanup the previous state. We'll
+		// live with the leftover state, however the user should be advised to
+		// eventually clean this up.
+		if err := m.deleteOldRoutes(toDelete); err != nil {
+			version := "new"
+			if rule.Priority == linux_defaults.RulePriorityEgressv2 {
+				version = "original"
+			}
+
+			scopedLog := log.WithField("rule", rule)
+			scopedLog.WithError(err).WithField("routes", routes).
+				Warnf("Failed to cleanup after successfully migrating endpoint to %s ENI datapath. "+
+					"It is recommended that theses routes are cleaned up, as it is possible in the future "+
+					"to collide with another endpoint with the same IP.", version)
```

Comment:
How would the user actually perform the cleanup? By running `ip route del` manually to remove routes with priority 110 for example? It's maybe worth giving the user a hint here.